core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/identity/BasePermission.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2021 B2i Healthcare Pte Ltd, http://b2i.sg
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.b2international.snowowl.core.identity;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.util.Objects;
+
+import org.apache.commons.io.FilenameUtils;
+
+import com.google.common.base.MoreObjects;
+
+/**
+ * @since 8.0
+ */
+public abstract class BasePermission implements Permission {
+
+	private static final long serialVersionUID = 1L;
+	
+	private final String operation;
+	
+	private String permission;
+	
+	public BasePermission(String operation) {
+		this.operation = checkNotNull(operation, "Operation must be specified.");
+	}
+	
+	@Override
+	public final String getOperation() {
+		return operation;
+	}
+	
+	@Override
+	public final String getPermission() {
+		if (permission == null) {
```

Comment:
For thread safety reasons, this should be put into a memoizing `Supplier` instead.